# MYC
Scripts for analysis conducted in preparation of MYC project/SF3B3 manuscript. Unless otherwise noted conditions are NEG-DMSO (ND), dTAG-DMSO (dD), NEG-P300i (NP), dTAG-P300i (dP) Experiments included in this repo are:

220519_CUTandRUN: Myc, PolII, H3K27me3

220902_ATAC: late timepoint ATACseq, n=2, originally processed by DV, then reprocessed, analysed and compared in my scripts.

230420_TTseq: Late timepoint TTseq, n=4, nascent and total fractions

230427_PROseq: Late timepoint PROseq, more sequencing of 230406_PROseq, tech reps merged in this dir (w/230406_PRO-seq_LATE), n=3

230727_ChIP: SF3B1, SF3B3, P300 ChIP

230815_CUTandRUN: Pretreat, early and late timepoints of H3K4me3, H3K27me3, H3K27ac & PolII

230816_Nanopore_nascent_WEHI: full nascent nanopore experiment, n=3

231005_ChIP: p300 ChIP (D2X6N), A485, dCBP1

231106_ChIP: P300 ChIP

231110_ChIP: SF3A3, U2AF2, SF3B1, SF3B3 ChIP

231205_ChIP: Early timepoint P300

231222_TTseq: Early timepoint TTseq (nascent and total fractions)

240111_ChIP: PolII (bulk & neg/pos reactivation sorts)

240116_ChIP: pS2 & pS5 PolII ChIP (bulk & neg/pos reactivation sorts)

240228_PROseq_EARLY: 3h timepoint +/- each of SF3B1-, SF3B3- & SF3B5-dTAG, n=2

240307_ChIP: H3K27ac and H2BK20ac ChIPs

240411_RNAseq: B3-NEG and -dTAG, CEBBPKO & EP300 KO, n=4

240419_ChIP: H2BK16ac, H3K4me1, H3K4me3, H3K9ac, dros spike-in, n=1

240424_ChIP: H3K27ac repeat w/dros spike-in, n=1

240508_ChIP: CDK9, DSSO, EGS, JUN, KAT2A, TBP, n=1

240523_PROseq_EARLY: resequencing of early timepoint PROseq, size selected to reduce adapter contamination

240531_SLAMseq: SF3B1 v B3 v B5 +/- dTAG 4h

GSE150625: public K562 parental cell PROseq https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150625
