---
title: "240523_PROseq"
author: "Andrea"
date: '2024-05-27'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

fastq dir: /pipeline/Runs/NextSeq/240523_VH01624_102_AAFJTJTM5/ProjectFolders/Project_Dane-Vassiliadis/
project dir: /dawson_genomics/Projects/MYC/240523_PROseq_EARLY

# run processing as in late timepoint PROseq
```{bash}
mkdir fastqc
mkdir fastp
mkdir fastq_screen
mkdir bams
mkdir tdfs
mkdir logs
mkdir counts
mkdir session-info

for fq in /pipeline/Runs/NextSeq/240228_VH01624_77_AAF7YYCM5/ProjectFolders/Project_Dane-Vassiliadis/*/*_R1_001.fastq.gz
do
dname=`dirname $fq`
bname=`basename $fq _R1_001.fastq.gz`
sbatch scripts/align_Cornell_method.sbatch $fq ${dname}/${bname}_R2_001.fastq.gz
done

# when trimming has started, run multiqc on untrimmed fastqc
module load multiqc/1.8
cd fastqc
multiqc .
cd ..

# after processing is done get multiqc on everything
multiqc .
```

# also screen trimmed fastqs
```{bash}
mkdir trimmed_fastq_screen

for fq in fastp/*_R1.trim.fq.gz
do
sbatch scripts/fastq_screen.sbatch $fq
done

# get multiqc for this dir
module load multiqc/1.8
cd trimmed_fastq_screen
multiqc .
cd ..
```

# merge bams and make bigwigs and make DT plots from all
```{bash}
mkdir bigwigs
mkdir plots
mkdir results

sbatch scripts/mergeBamCov.sbatch bams/SF3B1-C6-new-NEG-3h_S1.sort.dedup.bam bams/SF3B1-C6-old-NEG-3h_S3.sort.dedup.bam SF3B1-NEG
sbatch scripts/mergeBamCov.sbatch bams/SF3B1-C6-new-V-3h_S2.sort.dedup.bam bams/SF3B1-C6-old-V-3h_S4.sort.dedup.bam SF3B1-V
sbatch scripts/mergeBamCov.sbatch bams/SF3B3-C3-NEG-3h_S5.sort.dedup.bam bams/SF3B3-C5-NEG-3h_S7.sort.dedup.bam SF3B3-NEG
sbatch scripts/mergeBamCov.sbatch bams/SF3B3-C3-V-3h_S6.sort.dedup.bam bams/SF3B3-C5-V-3h_S8.sort.dedup.bam SF3B3-V
sbatch scripts/mergeBamCov.sbatch bams/SF3B5-B6-NEG-3h_S9.sort.dedup.bam bams/SF3B5-D7-NEG-3h_S11.sort.dedup.bam SF3B5-NEG
sbatch scripts/mergeBamCov.sbatch bams/SF3B5-B6-V-3h_S10.sort.dedup.bam bams/SF3B5-D7-V-3h_S12.sort.dedup.bam SF3B5-V

sbatch scripts/DTMat_GeneRegion_reactGenes_merged.sbatch
sbatch scripts/DTMat_GeneRegion_pcGenes_merged.sbatch
```

# merge bams from technical replicates then dedup and index
```{bash}
for bam in bams/*merged.sort.bam
do 
sbatch scripts/merge_tech_reps.sbatch $bam
done
```

# make bio rep merged bigwigs from tech rep merged bams for further analysis
```{bash}
for bam in bams/*techmerged.sort.dedup.bam
do
sbatch scripts/BamCov.sbatch $bam
done
```

# also need merged tech reps for separate bio reps
```{bash}
## just reused same script from above, but changed to handle *sort.dedup.bam files
# rather than *merged.sort.bam files
for bam in bams/*[0-9].sort.dedup.bam
do 
sbatch scripts/merge_tech_reps.sbatch $bam
done

# make bigwigs from files produced above also
for bam in bams/*[0-9].techmerged.sort.dedup.bam
do
sbatch scripts/BamCov.sbatch $bam
done
```

